### PR TITLE
Introducing GH Actions workflow to run llama-recipes PyTest tests

### PR DESCRIPTION
**Pull request by maximgroshev**
# What does this PR do?

Introduction of GitHub Actions CI/CD workflow to run PyTest tests on generic Ubuntu 20.04 GHA hosted Runners.

## Feature/Issue validation/testing

Workflow is configured to be triggered on each PR submitted to be merged with *main* branch when following paths are affected:

- src/llama-recipes/configs/*.py
- src/llama-recipes/utils/*.py
- src/llama-recipes/datasets/*.py
- src/llama-recipes/data/*.py
- src/llama-recipes/*.py

Here is the most recent PyTest test run executed on llama-recipes fork:  https://github.com/maximgroshev/llama-recipes/actions/runs/7906759630/job/21582327220

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: .github/workflows/.github/workflows/ARC_llama_recipes_pytest_on_single_nvidia_gpu_node.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      - 'src/llama-recipes/utils/*.py'
+      - 'src/llama-recipes/*.py'
+
+  # triggers workflow manually for debugging purposes.      
+  workflow_dispatch:
+    inputs:
+      runner:
+        description: 'GHA Runner Scale Set label to run workflow on.'
+        required: true
+        default: gha-runner-scale-set-nvidia-single-gpu
```
Comment (HamidShojanazeri) [inline .github/workflows/.github/workflows/ARC_llama_recipes_pytest_on_single_nvidia_gpu_node.yaml:17]: @maximgroshev I wonder if for the pytest we would need to fall back to GPU/ I guess as the first step we are targeting CPU for Pytests?

Is this a reserved instance?

File: .github/workflows/.github/workflows/ARC_llama_recipes_pytest_on_single_nvidia_gpu_node.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
+name: "[ARC][GPU] llama-recipes Pytest tests on Self-Hosted GPU k8s Nodes."
+on:
+  pull_request:
+    branches:    
+      - 'main'
+    paths:
+      - 'src/llama-recipes/configs/*.py'
+      - 'src/llama-recipes/utils/*.py'
+      - 'src/llama-recipes/*.py'
```
Comment (HamidShojanazeri) [inline .github/workflows/.github/workflows/ARC_llama_recipes_pytest_on_single_nvidia_gpu_node.yaml:9]: wonder if we should cover `src/llama-recipes/data/*.py`  and `src/llama-recipes/dataset/*.py`as well? as we have set of tests for [datasets](https://github.com/facebookresearch/llama-recipes/tree/main/tests/datasets)?

**Review by HamidShojanazeri — COMMENTED** [review]
Thanks @maximgroshev for the PR left couple of comments

File: .github/workflows/.github/workflows/ARC_llama_recipes_pytest_on_single_nvidia_gpu_node.yaml (hunk 1 above)
Comment (maximgroshev) [reply .github/workflows/.github/workflows/ARC_llama_recipes_pytest_on_single_nvidia_gpu_node.yaml:17]: I agree on running Pytest test on CPU k8s nodes. Currently there are still couple of dependencies on cuda in existing tests that make them fail on CPUI runners: https://github.com/maximgroshev/llama-recipes/actions/runs/7903261159

**Review by HamidShojanazeri — APPROVED** [review]
Thanks @maximgroshev for the PR and offline discussions.

**Pull request merged**